[PluginStats] Fix and Add command to PluginStats

SirWant opened the pull request:
Fix Std. deviation invalid commands.
Add commands: minN (maxN), sample, sampleN, size.
minN (maxN) - Compute min (max) over the last X samples (or less if there are less samples available).
sampleN - Compute X sample (sample1 - last (current)). If N>Number of saved samples - return error value(-1).
sample - The first (oldest) sample.
size - Number of saved samples.

SirWant committed:
Fix and Add command to PluginStats

Fix Std. deviation invalid commands
Add commands: minN (maxN), sampleN, size

SirWant committed:
Merge branch 'mega' into SirWant-patch-PluginStats

TD-er commented:
Looks good.
Only missing is the documentation :)

SirWant commented:
Sorry, I don't speak English (I use Google translator), it's better if someone else writes the documentation )

TD-er commented:
You don't need to 'speak' it, only writing it :)
But I get your point.

TD-er committed:
[Cleanup] Remove duplicate code & document

SirWant commented on docs/source/Plugin/_Plugin.rst:
```diff
@@ -97,6 +97,9 @@ For example using just like normal task value data:
 * ``[bme#temp.max]`` Refer to the maximum recorded sample since the last ``resetpeaks``. N.B. Not all tasks log the min and max peaks.
 * ``[bme#temp.min]`` See ``[bme#temp.max]`` 
+* ``[bme#temp.size]`` Return the number of samples in memory.
+* ``[bme#temp.sample]`` Access the last sample in memory.
```
not last, first (oldest) because last = sample1

TD-er replied:
I did change the behavior as it seemed a bit strange when considering the other use of "N".
If you also need the "oldest", maybe we can also add "first"?

SirWant replied:
Why different? It is different (reverse), from the point of view of the programmer, but not different for the user. This is clearly seen on the graph - the last X is counted from the end. The user is not comfortable counting from the middle. And then sample1, sample2, ... sample(last). Perhaps I initially did not correctly name - the first one is Unix time, but the last one is from the current time )

TD-er replied:
Hmm you might be on to something, like I messed up by using "0" instead of "size() - 1" or something like that.

But in order to grasp it, I really need some coffee (and a lot less interrupts which I'm now dealing with here)...
Will get back to this...

TD-er replied:
N.B. I really appreciate these kinds of "close read" reviews, so keep being critical :)

SirWant replied:
> "size() - 1"

Yes, if we reverse my logic, and sample = current, then i = lastNrSamples - 1.
Both methods are equal ) Without knowing the size, in one we can always get the penultimate sample, in the other the second, etc. You might think, maybe you can do both options, like sample-N
UPD: rather "-sample" and "-sampleN"

SirWant commented on src/src/DataStructs/PluginStats.cpp:
```diff
@@ -156,80 +167,64 @@ bool PluginStats::plugin_get_config_value_base(struct EventStruct *event, String
     success = true;
-  } else if (command.startsWith(F("sample"))) {
-    if (equals(command, F("sample"))) { // [taskname#valuename.sample] The first (oldest) sample saved.
-      value   = _samples[0];
```
value = getSample(_samples.size()) or value   = _samples[0]

SirWant commented on src/src/DataStructs/PluginStats.cpp:
```diff
@@ -156,80 +167,64 @@ bool PluginStats::plugin_get_config_value_base(struct EventStruct *event, String
-          success = true;
-        }
+    if (nrSamples < 0) {   // [taskname#valuename.sample] The last sample saved.
+      value = getSample(0);
```
value = getSample(_samples.size()) or value = _samples[0]

TD-er replied:
See my other comment.
And these kinds of comments from you are the reason I did not yet merge it :)

SirWant committed:
Replacing sampleN and sample to sample+-N

sample1 - last (current) sample
sample-1 - first of saved sample
sample - Number of saved samples (size command can be removed)

SirWant committed:
minor fix

SirWant committed:
Final minor fix...

TD-er commented on src/src/DataStructs/PluginStats.cpp:
```diff
@@ -120,14 +120,16 @@ float PluginStats::getSampleExtreme(PluginStatsBuffer_t::index_t lastNrSamples,
+  if (lastNrSamples >= 0) {
     i = _samples.size() - lastNrSamples;
+  } else {
+    i = abs(lastNrSamples) - 1;
```
Why use abs() here? Isn't that confusing to have negative indices?

SirWant replied:
No, I'm not sure that the code will be easier with negative indices. But I'm not a programmer )

TD-er commented on src/src/DataStructs/PluginStats.cpp:
```diff
@@ -120,14 +120,16 @@ float PluginStats::getSampleExtreme(PluginStatsBuffer_t::index_t lastNrSamples,
 
-  if (lastNrSamples < _samples.size()) {
+  if (lastNrSamples >= 0) {
     i = _samples.size() - lastNrSamples;
```
with the index being 0, you will get the wrong index.

SirWant replied:
This has already been fixed in the next commits, sorry.

SirWant committed:
Merge branch 'letscontrolit:mega' into SirWant-patch-PluginStats

TD-er committed:
Merge branch 'mega' into SirWant-patch-PluginStats
